beacon-chain/rpc/eth/v1/validator/validator.go

```diff
@@ -112,7 +95,60 @@ func (vs *Server) GetAttesterDuties(ctx context.Context, req *v1.AttesterDutiesR
 
 // GetProposerDuties requests beacon node to provide all validators that are scheduled to propose a block in the given epoch.
 func (vs *Server) GetProposerDuties(ctx context.Context, req *v1.ProposerDutiesRequest) (*v1.ProposerDutiesResponse, error) {
-	return nil, errors.New("Unimplemented")
+	ctx, span := trace.StartSpan(ctx, "validatorv1.GetProposerDuties")
+	defer span.End()
+
+	if vs.SyncChecker.Syncing() {
+		return nil, status.Error(codes.Unavailable, "Syncing to latest head, not ready to respond")
+	}
+
+	cs := vs.TimeFetcher.CurrentSlot()
+	currentEpoch := helpers.SlotToEpoch(cs)
+	if req.Epoch > currentEpoch {
+		return nil, status.Errorf(codes.InvalidArgument, "Request epoch %d can not be greater than current epoch %d", req.Epoch, currentEpoch)
+	}
+
+	s, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get head state: %v", err)
+	}
+
+	s, err = advanceState(ctx, s, req.Epoch, currentEpoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not advance state to requested epoch start slot: %v", err)
+	}
+
+	_, proposals, err := helpers.CommitteeAssignments(s, req.Epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not compute committee assignments: %v", err)
+	}
+
+	duties := make([]*v1.ProposerDuty, 0)
+	for index, slots := range proposals {
+		val, err := s.ValidatorAtIndexReadOnly(index)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get validator: %v", err)
+		}
+		pubkey48 := val.PublicKey()
+		pubkey := pubkey48[:]
+		for _, s := range slots {
+			duties = append(duties, &v1.ProposerDuty{
+				Pubkey:         pubkey,
+				ValidatorIndex: index,
+				Slot:           s,
```

Comment:
Do you know if `duties` shall be sorted by `slot` at the end?

Reply:
There is no such recommendation, although it seems the most sensible approach. I will sort them.